cpp/src/hdbscan/detail/reachability.cuh

```diff
@@ -162,6 +163,66 @@ void _compute_core_dists(const raft::handle_t& handle,
   core_distances<value_idx>(dists.data(), min_samples, min_samples, m, core_dists, stream);
 }
 
+//  Functor to post-process distances into reachability space
+template <typename value_idx, typename value_t>
+struct ReachabilityPostProcess {
+  DI value_t operator()(value_t value, value_idx row, value_idx col) const
+  {
+    return max(core_dists[col], max(core_dists[row], alpha * value));
+  }
+
+  const value_t* core_dists;
+  value_t alpha;
+};
+
+/**
+ * Given core distances, Fuses computations of L2 distances between all
+ * points, projection into mutual reachability space, and k-selection.
+ * @tparam value_idx
+ * @tparam value_t
+ * @param[in] handle raft handle for resource reuse
+ * @param[out] out_inds  output indices array (size m * k)
+ * @param[out] out_dists output distances array (size m * k)
+ * @param[in] X input data points (size m * n)
+ * @param[in] m number of rows in X
+ * @param[in] n number of columns in X
+ * @param[in] k neighborhood size (includes self-loop)
+ * @param[in] core_dists array of core distances (size m)
+ */
+template <typename value_idx, typename value_t>
+void mutual_reachability_knn_l2(const raft::handle_t& handle,
```

Comment:
The beauty of this is that it’s no longer just limited to l2 distance. What’s amazing about this is that it allows us to now use this epilogue with all of our distances! 

(Though perhaps we might need to be careful about how the fused l2 comes into play here? Maybe we need to wrap that epilogue internally or something. 

Since we are pushing up against release, maybe let’s push opening up the other distances until 23.06 or later. Can you create a cuml issue for that just to keep it on the radar and maybe reference it in the code here?

Reply:
sounds good! I created https://github.com/rapidsai/cuml/issues/5301 , and added a note in the last commit